### PR TITLE
feat: Locomotion

### PR DESCRIPTION
**Pull request by Kinerius**
# Roadmap

**Must Have**

- [x] Gravity
- [x] Walking / Jogging / Running
- [x] Coyote time
- [x] Jumping
- [x] Jump Buffering
- [x] Jump Grace Time
- [x] Falling / Hard Falling / Hard Landing
- [x] Step Handling
- [x] Moving Platforms
- [x] Camera Effects
- [x] Animations
- [x] Mechanim Integration
- [ ] Playground

**Nice to have**

- [x] Slope Management
- [x] IK: Foot Placement
- [ ] Wall Sliding
- [ ] Wall Stomp
- [ ] Keyboard Mapping
- [ ] Controller Integration
- [ ] NFT hook
- [ ] SDK Integration
- [x] Developer Tools

**Expendables**

- [ ] Sounds
- [ ] Particle Effects
- [x] IK: Look At
- [ ] Stamina

### CODE REVIEW

File: Explorer/Assets/DCL/AvatarRendering/AvatarShape/AvatarBase.cs [hunk 1]
```diff
@@ -19,15 +19,12 @@ public class AvatarBase : MonoBehaviour
         public void SetAsMainPlayer(bool isMainPlayer)
```
Comment (mikhail-dcl) [inline Explorer/Assets/DCL/AvatarRendering/AvatarShape/AvatarBase.cs:19]: This piece was a pure draft, this feature is the right moment to change it. I don't mind if it is done a bit later but by the end of this feature.

File: Explorer/Assets/DCL/CharacterMotion/Animation/ApplyJumpState.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace DCL.CharacterMotion.Animation
+{
+    public static class ApplyJumpState
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void Execute(ref CharacterAnimationComponent animationComponent,
+            in ICharacterControllerSettings settings,
+            in CharacterRigidTransform rigidTransform,
+            in AvatarBase avatarBase,
```
Comment (mikhail-dcl) [inline Explorer/Assets/DCL/CharacterMotion/Animation/ApplyJumpState.cs:14]: I propose to create an interface for `AvatarBase.avatarAnimator` to enable Testability. And in both this method and `ApplyMovementBlend` you don't really need to couple with `MonoBeahviours` (especially considering `AvatarBase` is a draft

File: Explorer/Assets/DCL/CharacterMotion/Components/CharacterRigidTransform.cs [hunk 3]
```diff
@@ -14,29 +14,53 @@ namespace DCL.CharacterMotion.Components
     /// </summary>
     public class CharacterRigidTransform
     {
+        public struct MovementVelocity
```
Comment (mikhail-dcl) [inline Explorer/Assets/DCL/CharacterMotion/Components/CharacterRigidTransform.cs:17]: Can you check settings of your auto clean-up on commit? Mine changes the layout and removes redundant namespaces

File: Explorer/Assets/DCL/CharacterMotion/Platforms/SaveLocalPosition.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
+﻿using DCL.CharacterMotion.Components;
+using UnityEngine;
+
+namespace DCL.CharacterMotion.Platforms
+{
+    public static class SaveLocalPosition
+    {
+        public static void Execute(ref CharacterPlatformComponent platformComponent, Vector3 position)
```
Comment (mikhail-dcl) [inline Explorer/Assets/DCL/CharacterMotion/Platforms/SaveLocalPosition.cs:8]: I'd add `AggressiveInlining` hint just like you did for some other methods

File: Explorer/Assets/DCL/CharacterMotion/Platforms/SaveLocalRotation.cs [hunk 5]
```diff
@@ -0,0 +1,15 @@
+﻿using DCL.CharacterMotion.Components;
+using UnityEngine;
+
+namespace DCL.CharacterMotion.Platforms
+{
+    public class SaveLocalRotation
+    {
+        public static void Execute(ref CharacterPlatformComponent platformComponent, Vector3 forward)
```
Comment (mikhail-dcl) [inline Explorer/Assets/DCL/CharacterMotion/Platforms/SaveLocalRotation.cs:8]: Here as well

File: Explorer/Assets/DCL/CharacterMotion/Platforms/SaveLocalRotation.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
+﻿using DCL.CharacterMotion.Components;
+using UnityEngine;
+
+namespace DCL.CharacterMotion.Platforms
+{
+    public class SaveLocalRotation
```
Comment (mikhail-dcl) [inline Explorer/Assets/DCL/CharacterMotion/Platforms/SaveLocalRotation.cs:6]: You can make this class static

File: Explorer/Assets/DCL/CharacterMotion/Animation/ApplyJumpState.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+            in StunComponent stunComponent)
+        {
+            var animator = avatarBase.avatarAnimator;
+
+            animationComponent.States.IsGrounded = rigidTransform.IsGrounded;
+            animationComponent.States.IsFalling = !rigidTransform.IsGrounded && rigidTransform.NonInterpolatedVelocity.y < -5f;
```
Comment (mikhail-dcl) [inline Explorer/Assets/DCL/CharacterMotion/Animation/ApplyJumpState.cs:20]: I believe we can move these magic constants to the settings

File: Explorer/Assets/DCL/CharacterMotion/Systems/CharacterPlatformSystem.cs [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void CheckPlatform(CharacterPlatformComponent platformComponent, Transform transform)
+        {
+            var ray = new Ray
+            {
+                origin = transform.position + (Vector3.up * 0.15f),
+                direction = Vector3.down,
+            };
+
+            bool hasHit = Physics.Raycast(ray, out RaycastHit hitInfo, 0.30f, PhysicsLayers.CHARACTER_ONLY_MASK);
```
Comment (mikhail-dcl) [inline Explorer/Assets/DCL/CharacterMotion/Systems/CharacterPlatformSystem.cs:64]: Magic constant

File: Explorer/Assets/DCL/CharacterMotion/Systems/CharacterPlatformSystem.cs [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+            characterController.Move(deltaPosition);
+            transform.forward = newCharacterForward;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void CheckPlatform(CharacterPlatformComponent platformComponent, Transform transform)
+        {
+            var ray = new Ray
+            {
+                origin = transform.position + (Vector3.up * 0.15f),
```
Comment (mikhail-dcl) [inline Explorer/Assets/DCL/CharacterMotion/Systems/CharacterPlatformSystem.cs:60]: Magic constant

File: Explorer/Assets/DCL/CharacterMotion/Systems/RotateCharacterSystem.cs [hunk 10]
```diff
@@ -15,46 +16,49 @@ namespace DCL.CharacterMotion.Systems
...
+    [UpdateInGroup(typeof(PresentationSystemGroup))]
     [UpdateAfter(typeof(CameraGroup))]
+    [UpdateAfter(typeof(InterpolateCharacterSystem))]
     public partial class RotateCharacterSystem : BaseUnityLoopSystem
     {
-        private SingleInstanceEntity camera;
-
         internal RotateCharacterSystem(World world) : base(world) { }
 
         public override void Initialize()
```
Comment (mikhail-dcl) [inline Explorer/Assets/DCL/CharacterMotion/Systems/RotateCharacterSystem.cs:27]: Let's remove an empty method

File: Explorer/Assets/DCL/CharacterMotion/Velocity/ApplyAirDrag.cs [hunk 11]
```diff
@@ -12,7 +13,25 @@ public static class ApplyAirDrag
...
+                tempVelocity.y = 0;
+                tempVelocity = ApplyDrag(tempVelocity, characterControllerSettings.AirDrag * characterControllerSettings.JumpVelocityDrag, deltaTime);
+                tempVelocity.y = characterPhysics.MoveVelocity.Velocity.y;
+
+                characterPhysics.MoveVelocity.Velocity = tempVelocity;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static Vector3 ApplyDrag(Vector3 vector, float drag, float deltaTime)
```
Comment (mikhail-dcl) [inline Explorer/Assets/DCL/CharacterMotion/Velocity/ApplyAirDrag.cs:27]: Don't we want to cover this logic with a Unit Test?

File: Explorer/Assets/DCL/CharacterMotion/Velocity/ApplySlopeModifier.cs [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+            position.y = feet;
+
+            // Todo: cache
+            var ray = new Ray
+            {
+                origin = position + (rigidTransform.MoveVelocity.Velocity * dt),
+                direction = Vector3.down,
+            };
+
+            float downwardsSlopeDistance = input.Kind == MovementKind.Run ? 0.55f : 0.45f;
```
Comment (mikhail-dcl) [inline Explorer/Assets/DCL/CharacterMotion/Velocity/ApplySlopeModifier.cs:30]: Magic constant

File: Explorer/Assets/DCL/CharacterMotion/Velocity/ApplySlopeModifier.cs [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+        {
+            // disabled when jumping or not grounded
+            if (!rigidTransform.IsGrounded || jump.IsPressed)
+                return Vector3.zero;
+
+            Vector3 position = characterController.transform.position;
+            float feet = position.y;
+            position.y = feet;
+
+            // Todo: cache
```
Comment (mikhail-dcl) [inline Explorer/Assets/DCL/CharacterMotion/Velocity/ApplySlopeModifier.cs:23]: What did you want to cache here?

File: Explorer/Assets/DCL/CharacterMotion/Velocity/ApplyVelocityStun.cs [hunk 14]
```diff
@@ -0,0 +1,14 @@
+﻿using DCL.CharacterMotion.Components;
+using UnityEngine;
+
+namespace DCL.CharacterMotion
+{
+    public static class ApplyVelocityStun
+    {
+        public static void Execute(ref CharacterRigidTransform rigidTransform, in StunComponent stunComponent)
```
Comment (mikhail-dcl) [inline Explorer/Assets/DCL/CharacterMotion/Velocity/ApplyVelocityStun.cs:8]: Aggressive Inlining

File: Explorer/Assets/DCL/Input/UIRaycaster.cs [hunk 15]
```diff
@@ -20,6 +20,17 @@ public UIRaycaster(EventSystem eventSystem)
         public IReadOnlyList<RaycastResult> RaycastAll(Vector2 position)
         {
             pointerEventData.position = position;
+            raycastResults.Clear();
+
+            bool isInSafeAreaWidth = position.x < 15 || position.x > Screen.width - 15;
```
Comment (mikhail-dcl) [inline Explorer/Assets/DCL/Input/UIRaycaster.cs:25]: Can we give the constant a name?

File: Explorer/Assets/DCL/Time/Systems/UpdateTimeSystem.cs [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+        {
+            UpdateTickQuery(World, t);
+        }
+
+        [Query]
+        private void UpdateTick(
+            [Data] float dt,
+            ref TimeComponent tickComponent)
+        {
+            tickComponent.Time += dt;
```
Comment (mikhail-dcl) [inline Explorer/Assets/DCL/Time/Systems/UpdateTimeSystem.cs:28]: I believe `deltaTime` is capped at a certain value so it's safer just to assign `UnityEngine.Time.time`

File: Explorer/Assets/Scripts/ECS/Abstract/SingleInstanceEntity.cs [hunk 17]
```diff
@@ -13,7 +14,12 @@ namespace ECS.Abstract
 
         public SingleInstanceEntity(in QueryDescription query, World world)
         {
+            TEMP[0] = Entity.Null;
             world.GetEntities(in query, TEMP);
+
+            if (TEMP[0] == Entity.Null)
+                throw new Exception("Entity not found");
```
Comment (mikhail-dcl) [inline Explorer/Assets/Scripts/ECS/Abstract/SingleInstanceEntity.cs:21]: Can we add information about `query` to this Exception?

**Review by mikhail-dcl — CHANGES_REQUESTED** [review]
The implementation looks gorgeous: very solid and clean 💪 
I just have several suggestions for further clean-up and tightening up

File: Explorer/Assets/DCL/CharacterMotion/Velocity/ApplySlopeModifier.cs (hunk 13 above)
Comment (Kinerius) [reply Explorer/Assets/DCL/CharacterMotion/Velocity/ApplySlopeModifier.cs:23]: Probably an old comment

**Review by mikhail-dcl — APPROVED** [review]
Thank you, all concerns were addressed ❤️

**Pull request merged**